modulus/datapipes/climate/climate.py

```diff
@@ -0,0 +1,790 @@
+        start_year: int,
+        num_samples_per_year: int,
+        latlon: np.ndarray,
+        variables: Union[List[str], None] = None,
+        aux_variables: List[Union[str,Callable]] = (),
+        batch_size: int = 1,
+        shuffle: bool = True,
+        process_rank: int = 0,
+        world_size: int = 1,
+    ):
+        self.data_paths = list(data_paths)
+        # Will be populated later once each worker starts running in its own process.
+        self.data_files = [None] * len(self.data_paths)
+        self.num_samples = num_samples
+        self.chans = list(channels)
+        self.latlon = latlon
+        self.variables = variables
+        self.aux_variables = aux_variables
+        self.num_steps = num_steps
+        self.stride = stride
+        self.dt = dt
+        self.start_year = start_year
+        self.num_samples_per_year = num_samples_per_year
+        self.batch_size = batch_size
+        self.shuffle = shuffle
+
+        self.last_epoch = None
+
+        self.indices = np.arange(num_samples)
+        # Shard from indices if running in parallel
+        self.indices = np.array_split(self.indices, world_size)[process_rank]
+
+        # Get number of full batches, ignore possible last incomplete batch for now.
+        # Also, DALI external source does not support incomplete batches in parallel mode.
+        self.num_batches = len(self.indices) // self.batch_size
+
+    @abstractmethod
+    def _load_sequence(self, year_idx: int, idx: int) -> np.array:
+        """Write data from year index `year_idx` and sample index `idx` to output"""
+        pass
+
+    def __call__(self, sample_info: dali.types.SampleInfo) -> Tuple[Tensor, np.ndarray]:
+        if sample_info.iteration >= self.num_batches:
+            raise StopIteration()
+
+        # Shuffle before the next epoch starts
+        if self.shuffle and sample_info.epoch_idx != self.last_epoch:
+            # All workers use the same rng seed so the resulting
+            # indices are the same across workers
+            np.random.default_rng(seed=sample_info.epoch_idx).shuffle(self.indices)
+            self.last_epoch = sample_info.epoch_idx
+
+        # Get local indices from global index
+        # TODO: This is very hacky, but it works for now
+        idx = self.indices[sample_info.idx_in_epoch]
+        year_idx = idx // self.num_samples_per_year
+        in_idx = idx % self.num_samples_per_year
+
+        state_seq = self._load_sequence(year_idx, in_idx)
+
+        # Load sequence of timestamps
+        year = self.start_year + year_idx
+        start_time = datetime(year, 1, 1) + timedelta(hours=int(in_idx) * self.dt)
+        timestamps = np.array(
+            [
+                (start_time + timedelta(hours=i * self.stride * self.dt)).timestamp()
+                for i in range(self.num_steps)
+            ]
+        )
+
+        # outputs from auxiliary sources
+        aux_outputs = (
+            callback(timestamps, self.latlon) 
```

Comment:
I'm a little confused here. It makes sense to give arbitrary functions of time and lat lon but when I look at the invariant functions they are only of lat lon and not with time. Am I missing something?

Reply:
Maybe it's a bit awkward, but the invariants that don't depend on time (like land-sea mask - hopefully sea level rise won't be so much that we have to reconsider this!) are set in the `invariants` argument of the datapipe, and these only accept the `latlon` argument. Meanwhile the time-dependent ones are set in `aux_variables` of the ClimateDataSourceSpec. This is because multiple time steps can be returned if `num_steps > 1`, in which case we also return the aux_variables for each time step.

Reply:
Oh I see. Makes sense